File: frontend/src/events/events.ts
import { Resource } from "../types/resource"
import Vector, { IVector } from "../types/vector"

export enum EVENT_TYPES {
    NEW_USER_EVENT = 0,
    USER_INIT_EVENT = 1,
    USER_MOVE_EVENT = 2,
    KEYBOARD_EVENT = 3,
    PLAYER_TARGET_POSITION_EVENT = 4,
    RESOURCE_POSITIONS_EVENT = 5,
    REMOVE_PLAYER_EVENT = 6,
    HIT_RESOURCE_EVENT = 7,
    UPDATE_RESOURCE_EVENT = 8,
    LOOT_RESOURCE_EVENT = 9,
    PLAYER_PLACED_RESOURCE_EVENT = 10,
    LOAD_INVENTORY_EVENT = 11,
    UPDATE_INVENTORY_EVENT = 12,
    REMOVE_GRID_CELL = 13,
    MULTIPLE_EVENTS = 14,
    LOGIN_PLAYER_EVENT = 15,
    CELL_DATA_EVENT = 16,
    NPC_LIST_EVENT = 17,
    NPC_TARGET_POSITION_EVENT = 18,
    HIT_NPC_EVENT = 19,
    UPDATE_NPC_EVENT = 20,
    UPDATE_PLAYER_EVENT = 21,
    NPC_ATTACK_ANIM_EVENT = 22,
}

export function createVector(x: number, y: number): Vector {
    return new Vector(x, y)
}

export interface BaseEvent {
    eventType: number
}

export function isBaseEvent(value: any): value is BaseEvent {
    return (
        "eventType" in value
    )
}

export interface NewPlayerEvent extends BaseEvent {
    id: number
    pos: {
        x: number,
        y: number
    }
    hitpoints: Hitpoints
}

export function isNewPlayerEvent(value: any): value is NewPlayerEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.NEW_USER_EVENT
    )
}

export interface GameConfig {
    gridCellSize: number
    subCells: number
    playerStepSize: number
    subCellSize: number
}

export interface UserInitEvent extends BaseEvent {
    id: number,
    uuid: string,
    pos: {
        x: number,
        y: number
    }
    hitpoints: Hitpoints
    gameConfig: GameConfig
}

export function isUserInitEvent(value: any): value is UserInitEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.USER_INIT_EVENT
    )
}

export interface PlayerTargetPositionEvent extends BaseEvent {
    id: number
    pos: Vector
    force: boolean
}

export function isPlayerTargetPositionEvent(value: any): value is PlayerTargetPositionEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.PLAYER_TARGET_POSITION_EVENT
    )
}

export interface RemovePlayerEvent extends BaseEvent {
    id: number
}

export function isRemovePlayerEvent(value: any): value is RemovePlayerEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.REMOVE_PLAYER_EVENT
    )
}


export enum KeyStates {
    UP,
    DOWN
}

export interface Hitpoints {
    current: number,
    max: number
}

export interface ResourceMin {
    resourceType: string,
    quantity: number,
}

export interface IResource {
    resourceType: string,
    id: number,
    gridCellKey: string,
    quantity: number,
    pos: Vector,
    hitpoints: Hitpoints
    isSolid: boolean,
    isLootable: boolean,
}

export interface ResourcePositionsEvent extends BaseEvent {
    resources: IResource[]
}

export function isResourcePositionsEvent(value: any): value is ResourcePositionsEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.RESOURCE_POSITIONS_EVENT
    )
}

export interface UpdateResourceEvent {
    id: number
    gridCellKey: string
    hitpoints: Hitpoints
    remove: boolean
    damage: number
}

export function isUpdateResourceEvent(value: any): value is UpdateResourceEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.UPDATE_RESOURCE_EVENT
    )
}

export interface UpdateNpcEvent {
    npcUUID: string
    gridCellKey: string
    hitpoints: Hitpoints
    remove: boolean
    damage: number
}

export function isUpdateNpcEvent(value: any): value is UpdateNpcEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.UPDATE_NPC_EVENT
    )
}

export interface LoadInventoryEvent extends BaseEvent {
    resources: ResourceMin[]
}

export function isLoadInventoryEvent(value: any): value is LoadInventoryEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.LOAD_INVENTORY_EVENT
    )
}

export interface UpdateInventoryEvent extends BaseEvent {
    resource: ResourceMin,
    remove: boolean
}

export function isUpdateInventoryEvent(value: any): value is UpdateInventoryEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.UPDATE_INVENTORY_EVENT
    )
}


export interface RemoveGridCellEvent extends BaseEvent {
    gridCellKey: string
}

export function isRemoveGridCellEvent(value: any): value is RemoveGridCellEvent {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.REMOVE_GRID_CELL
    )
}

export interface SubCell {
    pos: Vector,
    terrainType: string
}

export interface CellDataEvent extends BaseEvent {
    pos: {
        x: number,
        y: number
    },
    subCells: SubCell[]
    gridCellKey: string,
    subCellBase64: string
}

export function isCellDataEvent(value: any): value is CellDataEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.CELL_DATA_EVENT
    )
}

export interface INpc {
    UUID: string
    pos: IVector
    hitpoints: Hitpoints
    npcType: string
    attackSpeed: number
}

export interface NpcListEvent extends BaseEvent {
    gridCellKey: string
    npcList: INpc[]
}

export function isNpcListEvent(value: any): value is NpcListEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.NPC_LIST_EVENT
    )
}

export interface NpcTargetPositionEvent extends BaseEvent {
    gridCellKey: string,
    npcUUID: string
    pos: IVector
}

export function isNpcTargetPositionEvent(value: any): value is NpcTargetPositionEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.NPC_TARGET_POSITION_EVENT
    )
}

export interface UpdatePlayerEvent extends BaseEvent {
    playerId: number,
    hitpoints: Hitpoints,
    damage: number,
    heal: number,
    crit: boolean
}

export function isUpdatePlayerEvent(value: any): value is UpdatePlayerEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.UPDATE_PLAYER_EVENT
    )
}

export interface NpcAttackAnimEvent extends BaseEvent {
    npcUUID: string
    attackId: number
}

export function isNpcAttackAnimEvent(value: any): value is NpcAttackAnimEvent {
    return (
        isBaseEvent(value) &&
        value.eventType == EVENT_TYPES.NPC_ATTACK_ANIM_EVENT
    )
}

type EventTypes =
    NewPlayerEvent |
    UserInitEvent |
    PlayerTargetPositionEvent |
    RemovePlayerEvent |
    Hitpoints |
    IResource |
    ResourcePositionsEvent |
    UpdateResourceEvent |
    LoadInventoryEvent |
    UpdateInventoryEvent |
    RemoveGridCellEvent |
    MultipleEvents |
    CellDataEvent |
    NpcListEvent |
    NpcTargetPositionEvent |
    UpdatePlayerEvent |
    NpcAttackAnimEvent

export interface MultipleEvents extends BaseEvent {
    events: EventTypes[]
}

export function isMultipleEvents(value: any): value is MultipleEvents {
    return (
        isBaseEvent(value) &&
        value.eventType === EVENT_TYPES.MULTIPLE_EVENTS
    )
}


/*
 Events bellow are send from client to server 
*/


interface KeyBoardEvent extends BaseEvent {
    payload: {
        key: string,
        value: KeyStates,
    }
}

export function getKeyBoardEvent(key: string, value: KeyStates): string {
    const e: KeyBoardEvent = {
        "eventType": EVENT_TYPES.KEYBOARD_EVENT,
        "payload": {
            "key": key,
            "value": value,
        }
    }
    return JSON.stringify(e)
}

interface HitResourceEvent extends BaseEvent {
    payload: {
        skill: string,
        id: number,
    }
}

export function getHitResourceEvent(skill: string, id: number): string {
    const e: HitResourceEvent = {
        "eventType": EVENT_TYPES.HIT_RESOURCE_EVENT,
        "payload": {
            "skill": skill,
            "id": id,
        }
    }
    return JSON.stringify(e)
}

interface LootResourceEvent extends BaseEvent {
    payload: {
        id: number,
    }
}

export function getLootResourceEvent(resourceId: number): string {
    const e: LootResourceEvent = {
        "eventType": EVENT_TYPES.LOOT_RESOURCE_EVENT,
        "payload": {
            "id": resourceId,
        }
    }
    return JSON.stringify(e)
}

interface PlayerPlacedResourceEvent extends BaseEvent {
    payload: {
        resourceType: string,
        pos: Vector
    }
}

export function getPlayerPlacedResourceEvent(resourceType: string, pos: Vector): string {
    const e: PlayerPlacedResourceEvent = {
        "eventType": EVENT_TYPES.PLAYER_PLACED_RESOURCE_EVENT,
        "payload": {
            "resourceType": resourceType,
            "pos": pos
        }
    }
    return JSON.stringify(e)
}

interface LoginPlayerEvent extends BaseEvent {
    payload: {
        uuid: string,
    }
}

export function getLoginPlayerEvent(uuid: string): string {
    const e: LoginPlayerEvent = {
        "eventType": EVENT_TYPES.LOGIN_PLAYER_EVENT,
        "payload": {
            "uuid": uuid
        }
    }
    return JSON.stringify(e)
}

interface HitNpcEvent extends BaseEvent {
    payload: {
        skill: string,
        uuid: string,
    }
}

export function getHitNpcEvent(skill: string, uuid: string): string {
    const e: HitNpcEvent = {
        "eventType": EVENT_TYPES.HIT_NPC_EVENT,
        "payload": {
            "skill": skill,
            "uuid": uuid,
        }
    }
    return JSON.stringify(e)
}

File: server/root/events.go
package root

import (
	"encoding/json"
	"ws-game/item"
	"ws-game/resource"
	"ws-game/shared"
)

type EventType int

const (
	NEW_USER_EVENT               EventType = 0
	USER_INIT_EVENT              EventType = 1
	USER_MOVE_EVENT              EventType = 2
	KEYBOARD_EVENT               EventType = 3
	PLAYER_TARGET_POSITION_EVENT EventType = 4
	RESOURCE_POSITIONS_EVENT     EventType = 5
	REMOVE_PLAYER_EVENT          EventType = 6
	HIT_RESOURCE_EVENT           EventType = 7
	UPDATE_RESOURCE_EVENT        EventType = 8
	LOOT_RESOURCE_EVENT          EventType = 9
	PLAYER_PLACED_RESOURCE_EVENT EventType = 10
	LOAD_INVENTORY_EVENT         EventType = 11
	UPDATE_INVENTORY_EVENT       EventType = 12
	REMOVE_GRID_CELL             EventType = 13
	MULTIPLE_EVENTS              EventType = 14
	LOGIN_PLAYER_EVENT           EventType = 15
	CELL_DATA_EVENT              EventType = 16
	NPC_LIST_EVENT               EventType = 17
	NPC_TARGET_POSITION_EVENT    EventType = 18
	HIT_NPC_EVENT                EventType = 19
	UPDATE_NPC_EVENT             EventType = 20
	UPDATE_PLAYER_EVENT          EventType = 21
	NPC_ATTACK_ANIM_EVENT        EventType = 22
	ITEM_POSITIONS_EVENT         EventType = 23
	PLAYER_CLICKED_ITEM_EVENT    EventType = 24
	REMOVE_ITEM_EVENT            EventType = 25
	UPDATE_INVENTORY_ITEM_EVENT  EventType = 26
)

const (
	PLAYER_LOGIN_EVENT EventType = 15
)

type NewPlayerEvent struct {
	EventType EventType        `json:"eventType"`
	Id        int              `json:"id"`
	Pos       shared.Vector    `json:"pos"`
	Hitpoints shared.Hitpoints `json:"hitpoints"`
}

func GetNewPlayerEvent(id int, pos shared.Vector, hitpoints shared.Hitpoints) interface{} {
	return &NewPlayerEvent{EventType: NEW_USER_EVENT, Id: id, Pos: pos, Hitpoints: hitpoints}
}

type GameConfig struct {
	GridCellSize   int `json:"gridCellSize"`
	SubCells       int `json:"subCells"`
	PlayerStepSize int `json:"playerStepSize"`
	SubCellSize    int `json:"subCellSize"`
}

type UserInitEvent struct {
	EventType  EventType        `json:"eventType"`
	Id         int              `json:"id"`
	Pos        shared.Vector    `json:"pos"`
	Hitpoints  shared.Hitpoints `json:"hitpoints"`
	UUID       string           `json:"uuid"`
	GameConfig GameConfig       `json:"gameConfig"`
}

func NewUserInitEvent(id int, pos shared.Vector, hitpoints shared.Hitpoints, uuid string, config GameConfig) interface{} {
	return &UserInitEvent{
		EventType:  USER_INIT_EVENT,
		Id:         id,
		Pos:        pos,
		Hitpoints:  hitpoints,
		UUID:       uuid,
		GameConfig: config,
	}
}

type PlayerTargetPositionEvent struct {
	EventType EventType     `json:"eventType"`
	Id        int           `json:"id"`
	Pos       shared.Vector `json:"pos"`
	Force     bool          `json:"force"`
}

func NewPlayerTargetPositionEvent(v shared.Vector, id int, force bool) interface{} {
	return &PlayerTargetPositionEvent{EventType: PLAYER_TARGET_POSITION_EVENT, Pos: v, Id: id, Force: force}
}

type ResourcePositionsEvent struct {
	EventType EventType           `json:"eventType"`
	Resources []resource.Resource `json:"resources"`
}

func NewResourcePositionsEvent(resources map[int]resource.Resource) interface{} {

	v := make([]resource.Resource, 0, len(resources))

	for _, value := range resources {
		v = append(v, value)
	}

	return &ResourcePositionsEvent{EventType: RESOURCE_POSITIONS_EVENT, Resources: v}
}

type ItemPositionsEvent struct {
	EventType   EventType   `json:"eventType"`
	GridCellKey string      `json:"gridCellKey"`
	Items       []item.Item `json:"items"`
}

func NewItemPositionsEvent(items []item.Item, gridCellKey string) interface{} {

	return &ItemPositionsEvent{
		EventType:   ITEM_POSITIONS_EVENT,
		GridCellKey: gridCellKey,
		Items:       items,
	}
}

type RemoveItemEvent struct {
	EventType EventType `json:"eventType"`
	UUID      string    `json:"uuid"`
}

func NewRemoveItemEvent(uuid string) RemoveItemEvent {
	return RemoveItemEvent{EventType: REMOVE_ITEM_EVENT, UUID: uuid}
}

type NpcListEvent struct {
	EventType   EventType `json:"eventType"`
	GridCellKey string    `json:"gridCellKey"`
	NpcList     []Npc     `json:"npcList"`
}

func NewNpcListEvent(gridCellKey string, npcList []Npc) interface{} {
	return NpcListEvent{
		EventType:   NPC_LIST_EVENT,
		GridCellKey: gridCellKey,
		NpcList:     npcList,
	}
}

type RemovePlayerEvent struct {
	EventType EventType `json:"eventType"`
	Id        int       `json:"id"`
}

func NewRemovePlayerEvent(id int) interface{} {
	return &RemovePlayerEvent{EventType: REMOVE_PLAYER_EVENT, Id: id}
}

type LoadInventoryEvent struct {
	EventType EventType              `json:"eventType"`
	Resources []resource.ResourceMin `json:"resources"`
	Items     []item.Item            `json:"items"`
}

func NewLoadInventoryEvent(inventory map[resource.ResourceType]resource.Resource, items []item.Item) interface{} {

	resources := []resource.ResourceMin{}

	for _, entry := range inventory {
		resources = append(resources, resource.ResourceMin{
			Quantity:     entry.Quantity,
			ResourceType: entry.ResourceType,
		})
	}

	return &LoadInventoryEvent{
		EventType: LOAD_INVENTORY_EVENT,
		Resources: resources,
		Items:     items,
	}
}

type UpdateInventoryEvent struct {
	EventType EventType            `json:"eventType"`
	Resource  resource.ResourceMin `json:"resource"`
	Remove    bool                 `json:"remove"`
}

func NewUpdateInventoryEvent(r resource.ResourceMin, remove bool) interface{} {
	return &UpdateInventoryEvent{
		EventType: UPDATE_INVENTORY_EVENT,
		Resource:  r,
		Remove:    remove,
	}
}

type UpdateInventoryItemEvent struct {
	EventType EventType `json:"eventType"`
	Item      item.Item `json:"item"`
	Remove    bool      `json:"remove"`
}

func NewUpdateInventoryItemEvent(item item.Item, remove bool) interface{} {
	return &UpdateInventoryItemEvent{
		EventType: UPDATE_INVENTORY_ITEM_EVENT,
		Item:      item,
		Remove:    remove,
	}
}

type UpdateResourceEvent struct {
	EventType   EventType        `json:"eventType"`
	GridCellKey string           `json:"gridCellKey"`
	Id          int              `json:"id"`
	Hitpoints   shared.Hitpoints `json:"hitpoints"`
	Remove      bool             `json:"remove"`
	Damage      int              `json:"damage"`
}

func NewUpdateResourceEvent(id int, currentHp int, maxHp int, remove bool, gridCellKey string, damage int) interface{} {
	hitpoints := shared.Hitpoints{Current: currentHp, Max: maxHp}
	return &UpdateResourceEvent{EventType: UPDATE_RESOURCE_EVENT, Id: id, Remove: remove, Hitpoints: hitpoints, GridCellKey: gridCellKey, Damage: damage}
}

type RemoveGridCellEvent struct {
	EventType   EventType `json:"eventType"`
	GridCellKey string    `json:"gridCellKey"`
}

func NewRemoveGridCellEvent(gridCellKey string) interface{} {
	return &RemoveGridCellEvent{EventType: REMOVE_GRID_CELL, GridCellKey: gridCellKey}
}

type CellDataEvent struct {
	EventType     EventType     `json:"eventType"`
	GridCellKey   string        `json:"gridCellKey"`
	Pos           shared.Vector `json:"pos"`
	SubCells      []SubCell     `json:"subCells"`
	SubCellBase64 string        `json:"subCellBase64"`
}

func NewCellDataEvent(gridCellKey string, subCells []SubCell, pos shared.Vector, subCellBase64 string) interface{} {
	return &CellDataEvent{
		EventType:     CELL_DATA_EVENT,
		GridCellKey:   gridCellKey,
		SubCells:      subCells,
		Pos:           pos,
		SubCellBase64: subCellBase64,
	}
}

type MultipleEvents struct {
	EventType EventType     `json:"eventType"`
	Events    []interface{} `json:"events"`
}

func NewMultipleEvents(events []interface{}) interface{} {
	return &MultipleEvents{EventType: MULTIPLE_EVENTS, Events: events}
}

type NpcTagetPositionEvent struct {
	EventType   EventType     `json:"eventType"`
	GridCellKey string        `json:"gridCellKey"`
	NpcUUID     string        `json:"npcUUID"`
	Pos         shared.Vector `json:"pos"`
}

func NewNpcTargetPositionEvent(gridCellKey string, npcUUID string, pos shared.Vector) NpcTagetPositionEvent {
	return NpcTagetPositionEvent{
		EventType:   NPC_TARGET_POSITION_EVENT,
		GridCellKey: gridCellKey,
		NpcUUID:     npcUUID,
		Pos:         pos,
	}
}

type UpdateNpcEvent struct {
	EventType   EventType        `json:"eventType"`
	GridCellKey string           `json:"gridCellKey"`
	NpcUUID     string           `json:"npcUUID"`
	Hitpoints   shared.Hitpoints `json:"hitpoints"`
	Remove      bool             `json:"remove"`
	Damage      int              `json:"damage"`
}

func NewUpdateNpcEvent(uuid string, currentHp int, maxHp int, remove bool, gridCellKey string, damage int) interface{} {
	hitpoints := shared.Hitpoints{Current: currentHp, Max: maxHp}
	return &UpdateNpcEvent{EventType: UPDATE_NPC_EVENT, NpcUUID: uuid, Remove: remove, Hitpoints: hitpoints, GridCellKey: gridCellKey, Damage: damage}
}

type UpdatePlayerEvent struct {
	EventType EventType        `json:"eventType"`
	PlayerId  int              `json:"playerId"`
	Hitpoints shared.Hitpoints `json:"hitpoints"`
	Damage    int              `json:"damage"`
	Heal      int              `json:"heal"`
	Crit      bool             `json:"crit"`
}

func NewUpdatePlayerEvent(playerId int, hitpoints shared.Hitpoints, damage int, heal int, crit bool) interface{} {
	return &UpdatePlayerEvent{EventType: UPDATE_PLAYER_EVENT, PlayerId: playerId, Hitpoints: hitpoints, Damage: damage, Heal: heal, Crit: crit}
}

type NpcAttackAnimEvent struct {
	EventType EventType `json:"eventType"`
	NpcUUID   string    `json:"npcUUID"`
	AttackID  int       `json:"attackID"`
}

func NewNpcAttackAnimEvent(npcUUID string, attackID int) interface{} {
	return &NpcAttackAnimEvent{EventType: NPC_ATTACK_ANIM_EVENT, NpcUUID: npcUUID, AttackID: attackID}
}

// Events send from client

type BaseEvent struct {
	EventType EventType       `json:"eventType"`
	Payload   json.RawMessage `json:"payload"`
}
type KeyBoardEvent struct {
	Key   string `json:"key"`
	Value int    `json:"value"`
}

type HitResourceEvent struct {
	Skill string `json:"skill"`
	Id    int    `json:"id"`
}

type HitNpcEvent struct {
	Skill string `json:"skill"`
	UUID  string `json:"uuid"`
}

type LootResourceEvent struct {
	Id int `json:"id"`
}

type PlayerPlacedResourceEvent struct {
	ResourceType string        `json:"resourceType"`
	Pos          shared.Vector `json:"pos"`
}

type LoginPlayerEvent struct {
	ResourceType string `json:"resourceType"`
	UUID         string `json:"uuid"`
}

type PlayerClickedItemEvent struct {
	UUID        string        `json:"uuid"`
	GridCellPos shared.Vector `json:"gridCellPos"`
}
